### PR TITLE
Feature/jff 4/token request response

### PR DESCRIPTION
**Pull request by hrehman-sphereon**

### CODE REVIEW

File: src/types/OIDC4VCI.types.ts [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+export enum TokenType {
+  BEARER = 'bearer',
+}
+
+export enum ClientType {
+  CONFIDENTIAL = 'confidential',
+  PUBLIC = 'public',
+}
+
+export class ClientProfile {
```
[the remark this answers is not in the document]
Comment (hrehman-sphereon) [reply src/types/OIDC4VCI.types.ts]: This one has been removed.

File: README.md [hunk 2]
```diff
@@ -1,13 +1,16 @@
 <h1 align="center">
   <br>
   <a href="https://www.sphereon.com"><img src="https://sphereon.com/content/themes/sphereon/assets/img/logo.svg" alt="Sphereon" width="400"></a>
-    <br>OIDC4VCI-CLIENT 
-    <br>A client to request and receive Verifiable Credentials using Oidc4vci
+    <br>OpenIDConnect 4 Verifiable Credential Issuer - Client
```
Comment (nklomp) [inline README.md:4]:
```suggestion
    <br>OpenID for Verifiable Credential Issuance - Client 
```

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -25,14 +25,4 @@ jobs:
...
-        with:
-#          token: ${{ secrets.CODECOV_TOKEN }} # not required for public repos
-          name: codecove # optional
-          flags: unittest
-          fail_ci_if_error: true # optional (default = false)
-          version: "v0.1.15"
-          #directory: ./coverage/reports/
-          #files: ./coverage1.xml,./coverage2.xml
-          verbose: true # optional (default = false)
+
```
Comment (nklomp) [inline .github/workflows/main.yml:28]:
```suggestion
 uses: codecov/codecov-action@v2
        with:
#          token: ${{ secrets.CODECOV_TOKEN }} # not required for public repos
          name: codecove # optional
          flags: unittest
          verbose: true # optional (default = false)
```

File: README.md [hunk 4]
```diff
@@ -1,13 +1,16 @@
...
-    <br>A client to request and receive Verifiable Credentials using Oidc4vci
+    <br>OpenIDConnect 4 Verifiable Credential Issuer - Client 
   <br>
 </h1>
 
 _IMPORTANT it still in development and it's not fully functional_
 
+### Background 
+
+A client to request and receive Verifiable Credentials using Oidc4vci
```
Comment (nklomp) [inline README.md:12]:
```suggestion
A client to request and receive Verifiable Credentials using OID4VCI.
```

File: package.json [hunk 5]
```diff
@@ -18,10 +18,7 @@
     "test:lint": "eslint . --ext .ts",
     "test:prettier": "prettier \"{src,__test__}/**/*.ts\" --list-different",
     "test:cov": "jest --ci --coverage && codecov",
-    "uninstall": "rimraf dist coverage node_modules"
-  },
-  "engines": {
-    "node": ">=14"
+    "uninstall": "rimraf dist coverage yarn.lock package-lock.json node_modules"
   },
```
Comment (nklomp) [inline package.json:22]:
```suggestion
  },
  "engines": {
    "node": ">=14"
  },
```

File: src/AccessTokenClient.ts [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    }
+  }
+
+  private validate(authzFlowType: AuthzFlowType, accessTokenRequest: AccessTokenRequest, isPinRequired: boolean): void {
+    if (authzFlowType === AuthzFlowType.PRE_AUTHORIZED_CODE_FLOW) {
+      this.assertPreAuthorizedGrantType(accessTokenRequest.grant_type);
+      this.assertNonEmptyPreAuthorizedCode(accessTokenRequest);
+      this.assertNumericPin(accessTokenRequest, isPinRequired);
+      this.assertNonEmptyClientId(accessTokenRequest);
+    }
```
Comment (nklomp) [inline src/AccessTokenClient.ts:44]: Please add an else with a 'not supported yet' error, so we don't get questions about why the other flow doesn't work

File: src/functions/Encoding.ts [hunk 7]
```diff
@@ -1,57 +1,71 @@
...
+  function encodeAndStripWhitespace(key: string): string {
     return encodeURIComponent(key.replace(' ', ''));
   }
 
   for (const [key, value] of Object.entries(json)) {
     if (!value) {
       continue;
     }
+    //Skip properties that are not of URL type
+    if (!['issuer', 'credential_type'].includes(key)) {
```
Comment (nklomp) [inline src/functions/Encoding.ts:30]: Should be rebased on @maikel-maas' PR, as this shouldn't present anymore

File: src/types/OIDC4VCI.types.ts [hunk 8]
```diff
@@ -0,0 +1,55 @@
+export enum GrantTypes {
+  AUTHORIZATION_CODE = 'authorization_code',
+  PRE_AUTHORIZED = 'urn:ietf:params:oauth:grant-type:pre-authorized_code',
```
Comment (nklomp) [inline src/types/OIDC4VCI.types.ts:3]: PRE_AUTHORIZED_CODE

**Review by nklomp — CHANGES_REQUESTED** [review]
Some remarks/suggestions

File: src/types/OIDC4VCI.types.ts (hunk 8 above)
Comment (hrehman-sphereon) [reply src/types/OIDC4VCI.types.ts:3]: This is grant type ... This suggested name will be confused with the field that contains the code.

File: src/AccessTokenClient.ts (hunk 6 above)
Comment (hrehman-sphereon) [reply src/AccessTokenClient.ts:46]: fixed

File: src/functions/Encoding.ts (hunk 7 above)
Comment (hrehman-sphereon) [reply src/functions/Encoding.ts]: reabsed again.

File: src/AccessTokenClient.ts [hunk 9]
```diff
@@ -41,6 +41,8 @@ export class AccessTokenClient {
       this.assertNonEmptyPreAuthorizedCode(accessTokenRequest);
       this.assertNumericPin(accessTokenRequest, isPinRequired);
       this.assertNonEmptyClientId(accessTokenRequest);
+    } else {
+      throw new Error('Non-pre-authorized flow is not yet supported.');
```
Comment (nklomp) [inline src/AccessTokenClient.ts:45]: Loving you double negatives 😂 
'Only pre-authorized-code flow is supported' ;)

File: src/types/OIDC4VCI.types.ts (hunk 8 above)
Comment (hrehman-sphereon) [reply src/types/OIDC4VCI.types.ts]: PRE_AUTHORIZED_CODE and AUTHORIZATION_CODE are basically referring to the same thing. 
I believe when the grant is being given the server refers to that grant as 'AUTHORIZATION_CODE '.
When we are requesting access token we refer to our grant 'PRE_AUTHORIZED'.

So the server is sending us auth code. and we are telling server to give us the access code as we are pre-authroized.

That was the sense I got from it.

But pushed the fix.

Comment (hrehman-sphereon) [reply src/types/OIDC4VCI.types.ts]: And when were are creating the request ... we say that since our grant is 'pre-authorized' or auth-code is 'xyza' ...

Comment (nklomp) [reply src/types/OIDC4VCI.types.ts]: They are certainly not the same thing. Different code types for different flows

Comment (nklomp) [reply src/types/OIDC4VCI.types.ts]: BTW, you forgot to update the usage of the enum everywhere

**Comment by nklomp** [comment]
Build fails because of the updated enum name

File: src/AccessTokenClient.ts (hunk 9 above)
Comment (hrehman-sphereon) [reply src/AccessTokenClient.ts]: pushing fix.

File: src/types/OIDC4VCI.types.ts (hunk 8 above)
Comment (hrehman-sphereon) [reply src/types/OIDC4VCI.types.ts]: Refactor was indeed in the next commit.

Comment (hrehman-sphereon) [reply src/types/OIDC4VCI.types.ts]: I am keeping an eye on the build in git. I hope it will be passing and all comments resolved.

**Review by nklomp — APPROVED** [review]

**Comment by nklomp** [comment]
LGTM

**Pull request merged**